### PR TITLE
refactor:  Migrate to new Atlas SDK - mongodbatlas_federated_database_instance resource

### PR DESCRIPTION
**Pull request by maastha**
## Description

This PR:
1. Upgrade Atlas SDK version - go.mongodb.org/atlas-sdk/
2. Update/migrate `mongodbatlas_federated_database_instance` to use the new SDK
3. Add migration tests to test SDK update
4. Add deprecation message for `storage_stores.#.cluster_id` attribute in `mongodbatlas_federated_database_instance` resource and data source schemas

Link to any related issue(s): [INTMDB-1036](https://jira.mongodb.org/browse/INTMDB-1036)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [X] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

### CODE REVIEW

File: mongodbatlas/resource_mongodbatlas_federated_database_instance.go [hunk 1]
```diff
@@ -529,151 +517,127 @@ func newStores(d *schema.ResourceData) []*matlas.DataFederationStore {
...
+func newReadPreference(storeFromConfMap map[string]interface{}) *admin.DataLakeAtlasStoreReadPreference {
 	readPreferenceFromConf, ok := storeFromConfMap["read_preference"].([]interface{})
 	if !ok || len(readPreferenceFromConf) == 0 {
 		return nil
 	}
 	readPreferenceFromConfMap := readPreferenceFromConf[0].(map[string]interface{})
-	return &matlas.ReadPreference{
-		Mode:                readPreferenceFromConfMap["mode"].(string),
-		MaxStalenessSeconds: int32(readPreferenceFromConfMap["max_staleness_seconds"].(int)),
-		TagSets:             newTagSets(readPreferenceFromConfMap),
```
Comment (maastha) [inline mongodbatlas/resource_mongodbatlas_federated_database_instance.go:554]: Removing TagSets implementation because of difference in data type. Will be updated in separate PR.

File: mongodbatlas/resource_mongodbatlas_federated_database_instance.go [hunk 2]
```diff
@@ -239,10 +240,6 @@ func schemaFederatedDatabaseInstanceStores() *schema.Schema {
 					Type:     schema.TypeString,
 					Optional: true,
 				},
-				"cluster_id": {
```
Comment (maastha) [inline mongodbatlas/resource_mongodbatlas_federated_database_instance.go:242]: Removing cluster_id as this has been deprecated and not available in the new SDK anymore.

Comment (maastha) [reply mongodbatlas/resource_mongodbatlas_federated_database_instance.go:242]: cc @Zuhairahmed

File: mongodbatlas/resource_mongodbatlas_project.go [hunk 3]
```diff
@@ -8,12 +8,13 @@ import (
 	"net/http"
 	"time"
 
+	"go.mongodb.org/atlas-sdk/v20230201005/admin"
```
Comment (wtrocki) [inline mongodbatlas/resource_mongodbatlas_project.go]: nit: consistency with declaration of the dependencies

```suggestion
	atlasSDK "go.mongodb.org/atlas-sdk/v20230201005/admin"
```

**Comment by wtrocki** [comment]
LGTM

File: go.mod [hunk 4]
```diff
@@ -8,14 +8,15 @@ require (
...
+	github.com/hashicorp/terraform-plugin-log v0.9.0
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.27.0
 	github.com/hashicorp/terraform-plugin-testing v1.4.0
 	github.com/mongodb-forks/digest v1.0.5
 	github.com/mwielbut/pointy v1.1.0
 	github.com/spf13/cast v1.5.1
 	github.com/zclconf/go-cty v1.13.2
 	go.mongodb.org/atlas v0.33.0
-	go.mongodb.org/atlas-sdk/v20230201002 v20230201002.0.0
+	go.mongodb.org/atlas-sdk/v20230201005 v20230201005.0.1-0.20230825095504-8d6f690cf806
```
Comment (andreaangiolillo) [inline go.mod]: Make sure to add the latest release https://github.com/mongodb/atlas-sdk-go/releases/tag/v20230201006.0.0 before merging the PR

File: mongodbatlas/provider.go [hunk 5]
```diff
@@ -799,3 +800,17 @@ func expandAssumeRole(tfMap map[string]interface{}) *AssumeRole {
...
+	}
+	return nil
+}
+
+func stringPtr(v string) *string {
+	if v != "" {
+		return &v
+	}
+	return nil
+}
```
Comment (andreaangiolillo) [inline mongodbatlas/provider.go:817]: Why you need to define these function instead of using the one at line 800 (`pointer`)?

File: mongodbatlas/provider_test.go [hunk 6]
```diff
@@ -249,3 +255,19 @@ func testCheckPrivateEndpointServiceDataFederationOnlineArchiveRun(tb testing.TB
...
 }
+
+var _ plancheck.PlanCheck = debugPlan{}
+
+type debugPlan struct{}
+
+func (e debugPlan) CheckPlan(ctx context.Context, req plancheck.CheckPlanRequest, resp *plancheck.CheckPlanResponse) {
+	rd, err := json.Marshal(req.Plan)
+	if err != nil {
+		tflog.Debug(ctx, fmt.Sprintf("error marshaling machine-readable plan output: %s", err))
```
Comment (andreaangiolillo) [inline mongodbatlas/provider_test.go]: why don't we return an error here and block the execution?

File: mongodbatlas/resource_mongodbatlas_federated_database_instance.go [hunk 7]
```diff
@@ -529,151 +533,126 @@ func newStores(d *schema.ResourceData) []*matlas.DataFederationStore {
...
+			MaxWildcardCollections: intPtr(storageDBFromConfMap["max_wildcard_collections"].(int)),
 			Collections:            newDataFederationCollections(storageDBFromConfMap),
 		}
 	}
 
 	return dbs
 }
 
-func newDataFederationCollections(storageDBFromConfMap map[string]interface{}) []*matlas.DataFederationCollection {
+func newDataFederationCollections(storageDBFromConfMap map[string]interface{}) []admin.DataLakeDatabaseCollection {
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_federated_database_instance.go:586]:
```suggestion
func newDataFederationCollections(storageDBFromConfMap map[string]interface{}) []*admin.DataLakeDatabaseCollection {
```

File: mongodbatlas/resource_mongodbatlas_federated_database_instance.go [hunk 8]
```diff
@@ -529,151 +533,126 @@ func newStores(d *schema.ResourceData) []*matlas.DataFederationStore {
...
+			Name:        stringPtr(collectionFromConf.(map[string]interface{})["name"].(string)),
 			DataSources: newDataFederationDataSource(collectionFromConf.(map[string]interface{})),
 		}
 	}
 
 	return collections
 }
 
-func newDataFederationDataSource(collectionFromConf map[string]interface{}) []*matlas.DataFederationDataSource {
+func newDataFederationDataSource(collectionFromConf map[string]interface{}) []admin.DataLakeDatabaseDataSourceSettings {
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_federated_database_instance.go:603]:
```suggestion
func newDataFederationDataSource(collectionFromConf map[string]interface{}) []*admin.DataLakeDatabaseDataSourceSettings {
```

File: mongodbatlas/resource_mongodbatlas_federated_database_instance.go [hunk 9]
```diff
@@ -747,91 +728,89 @@ func flattenDataFederationDatabase(atlasDatabases []*matlas.DataFederationDataba
...
+			"source":   atlasView.GetSource(),
+			"pipeline": atlasView.GetPipeline(),
 		}
 	}
 
 	return views
 }
 
-func flattenDataFederationCollections(atlasCollections []*matlas.DataFederationCollection) []map[string]interface{} {
+func flattenDataFederationCollections(atlasCollections []admin.DataLakeDatabaseCollection) []map[string]interface{} {
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_federated_database_instance.go:750]:
```suggestion
func flattenDataFederationCollections(atlasCollections []*admin.DataLakeDatabaseCollection) []map[string]interface{} {
```

File: mongodbatlas/provider_test.go [hunk 10]
```diff
@@ -249,3 +255,19 @@ func testCheckPrivateEndpointServiceDataFederationOnlineArchiveRun(tb testing.TB
 		tb.Skip("`MONGODB_ATLAS_PRIVATE_ENDPOINT_ID` must be set for Private Endpoint Service Data Federation and Online Archive acceptance testing")
 	}
 }
+
+var _ plancheck.PlanCheck = debugPlan{}
+
+type debugPlan struct{}
+
+func (e debugPlan) CheckPlan(ctx context.Context, req plancheck.CheckPlanRequest, resp *plancheck.CheckPlanResponse) {
```
Comment (andreaangiolillo) [inline mongodbatlas/provider_test.go]: where is this function used?

File: mongodbatlas/resource_mongodbatlas_federated_database_instance_migration_test.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+import (
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/plancheck"
+)
+
+func TestAccFederatedDatabaseInstance_MigraitonTest_basic(t *testing.T) {
```
Comment (AgustinBettati) [inline mongodbatlas/resource_mongodbatlas_federated_database_instance_migration_test.go:14]: Could be worth leaving a comment adding context for the definition of these migration tests, because in this case it is not related to the framework migration but for migrating the SDK.

File: mongodbatlas/provider_test.go (hunk 10 above)
Comment (AgustinBettati) [inline mongodbatlas/provider_test.go]: we could extract these acceptance test utilities (including all the `testCheck*` functions) currently defined in `provider_test.go` into a separate package like `./mongodbatlas/acctest/utils.go`.

**Review by AgustinBettati — COMMENTED** [review]
LGTM, added some small comments

File: mongodbatlas/provider_test.go (hunk 10 above)
Comment (AgustinBettati) [reply mongodbatlas/provider_test.go]: Revisiting this comment. Can see that in migration branch we already created a `mongodbatlas/testutils` folder, so new file could be places here.

File: mongodbatlas/resource_mongodbatlas_federated_database_instance_migration_test.go (hunk 11 above)
Comment (maastha) [reply mongodbatlas/resource_mongodbatlas_federated_database_instance_migration_test.go:14]: @AgustinBettati True, but essentially we can be using the same tests for both migrations. This would help us to identify any breaking changes we might be introducing as part of either migration.

We can always extend the same tests if required during migration to the Framework.

File: mongodbatlas/provider.go (hunk 5 above)
Comment (maastha) [reply mongodbatlas/provider.go:817]: `pointer` function only checks whether the passed attribute is nil or not, we need to add additional checks here for 0s and empty strings because of the issue related to default values in Atlas and the workaround suggested by @AgustinBettati in Slack (shared). Essentially, we're not sending empty strings in the API requests.

Comment (andreaangiolillo) [reply mongodbatlas/provider.go:817]: thanks for the clarification

File: go.mod (hunk 4 above)
Comment (maastha) [reply go.mod]: updated

File: mongodbatlas/provider_test.go (hunk 6 above)
Comment (maastha) [reply mongodbatlas/provider_test.go]: This method is essentially used to ensure that the generated plan is empty, it is the same that we are using for migration to the Plugin Framework as well. Not sure I understand what you mean by "block the execution"?

File: mongodbatlas/resource_mongodbatlas_federated_database_instance.go (hunk 7 above)
Comment (maastha) [reply mongodbatlas/resource_mongodbatlas_federated_database_instance.go:586]: `[]admin.DataLakeDatabaseCollection` is the required data type by the SDK

File: mongodbatlas/resource_mongodbatlas_federated_database_instance.go (hunk 8 above)
Comment (maastha) [reply mongodbatlas/resource_mongodbatlas_federated_database_instance.go:603]: `[]admin.DataLakeDatabaseDataSourceSettings ` is the required data type by the SDK, unable to use pointers as above.

File: mongodbatlas/resource_mongodbatlas_federated_database_instance.go (hunk 9 above)
Comment (maastha) [reply mongodbatlas/resource_mongodbatlas_federated_database_instance.go:750]: `atlasCollections []admin.DataLakeDatabaseCollection` is the data type returned by the SDK, unable to use pointers as above

File: mongodbatlas/provider_test.go (hunk 10 above)
Comment (maastha) [reply mongodbatlas/provider_test.go]: moved to `mongodbatlas/testutils`

Comment (maastha) [reply mongodbatlas/provider_test.go]: This method is essentially used to ensure that the generated plan is empty, it is the same that we are using for migration to the Plugin Framework as well. 

I have added the same migration tests here that we use for migrating to the Plugin Framework as is, as the tests essentially help us identify if we might be introducing any unwanted changes to existing implementations of the resources.
Same tests can be re-used/extended for both migrations in the future.

File: mongodbatlas/resource_mongodbatlas_federated_database_instance_migration_test.go (hunk 11 above)
Comment (AgustinBettati) [reply mongodbatlas/resource_mongodbatlas_federated_database_instance_migration_test.go:14]: Makes sense 👍

**Review by AgustinBettati — APPROVED** [review]

**Review by andreaangiolillo — APPROVED** [review]
LGTM

File: mongodbatlas/resource_mongodbatlas_federated_database_instance.go [hunk 12]
```diff
@@ -240,8 +241,9 @@ func schemaFederatedDatabaseInstanceStores() *schema.Schema {
 					Optional: true,
 				},
 				"cluster_id": {
-					Type:     schema.TypeString,
-					Optional: true,
+					Type:       schema.TypeString,
+					Optional:   true,
+					Deprecated: fmt.Sprintf(DeprecationByDateMessageParameter, "September 2024"),
```
Comment (maastha) [inline mongodbatlas/resource_mongodbatlas_federated_database_instance.go:246]: @Zuhairahmed Could you please confirm if the deprecation message is okay for you?

Comment (Zuhairahmed) [reply mongodbatlas/resource_mongodbatlas_federated_database_instance.go:246]: LGTM

**Review by Zuhairahmed — APPROVED** [review]
